File: Dagger2Ex/app/src/main/java/expencive/vk/com/dagger2ex/dagger/WheelsModule.java
package expencive.vk.com.dagger2ex.dagger;


import dagger.Module;
import dagger.Provides;
import expencive.vk.com.dagger2ex.car.Rims;
import expencive.vk.com.dagger2ex.car.Tires;
import expencive.vk.com.dagger2ex.car.Wheels;

@Module
public class WheelsModule {

    @Provides
    static Rims provideRims() {
        return new Rims();
    }

    @Provides
    static Tires provideTires() {
        Tires tires = new Tires();
        tires.inflate();
        return tires;
    }

    @Provides
    static Wheels provideWheels(Rims rims, Tires tires) {
        return new Wheels(rims,tires);
    }
}

File: DaggerKotlin/app/src/main/java/vk/expencive/daggerkotlin/dagger/WheelsModule.kt
package vk.expencive.daggerkotlin.dagger

import dagger.Module
import dagger.Provides
import vk.expencive.daggerkotlin.car.Rims
import vk.expencive.daggerkotlin.car.Tires
import vk.expencive.daggerkotlin.car.Wheels

@Module
class WheelsModule {

    @Provides
    fun  provideRims(): Rims =
        Rims()

    @Provides
    fun provideTires(): Tires {
        var tires = Tires()
        tires.inflate()
        return tires;
    }

    @Provides
    fun provideWheels(rims: Rims, tires: Tires) =
        Wheels(rims, tires)


}